eclair-core/src/main/scala/fr/acinq/eclair/db/sqlite/SqliteAuditDb.scala

```diff
@@ -175,61 +208,94 @@ class SqliteAuditDb(sqlite: Connection) extends AuditDb with Logging {
       }
-      q
+      sentByParentId.values.toSeq.sortBy(_.timestamp)
     }
 
   override def listReceived(from: Long, to: Long): Seq[PaymentReceived] =
-    using(sqlite.prepareStatement("SELECT * FROM received WHERE timestamp >= ? AND timestamp < ? ORDER BY timestamp")) { statement =>
+    using(sqlite.prepareStatement("SELECT * FROM received WHERE timestamp >= ? AND timestamp < ?")) { statement =>
       statement.setLong(1, from)
       statement.setLong(2, to)
       val rs = statement.executeQuery()
-      var q: Queue[PaymentReceived] = Queue()
+      var receivedByHash = Map.empty[ByteVector32, PaymentReceived]
       while (rs.next()) {
-        q = q :+ PaymentReceived(
-          rs.getByteVector32("payment_hash"),
-          Seq(PaymentReceived.PartialPayment(
-            MilliSatoshi(rs.getLong("amount_msat")),
-            rs.getByteVector32("from_channel_id"),
-            rs.getLong("timestamp")
-          )))
+        val paymentHash = rs.getByteVector32("payment_hash")
+        val part = PaymentReceived.PartialPayment(
+          MilliSatoshi(rs.getLong("amount_msat")),
+          rs.getByteVector32("from_channel_id"),
+          rs.getLong("timestamp"))
+        val received = receivedByHash.get(paymentHash) match {
+          case Some(r) => r.copy(parts = r.parts :+ part)
+          case None => PaymentReceived(paymentHash, Seq(part))
+        }
+        receivedByHash = receivedByHash + (paymentHash -> received)
       }
-      q
+      receivedByHash.values.toSeq.sortBy(_.timestamp)
     }
 
   override def listRelayed(from: Long, to: Long): Seq[PaymentRelayed] =
-    using(sqlite.prepareStatement("SELECT * FROM relayed WHERE timestamp >= ? AND timestamp < ? ORDER BY timestamp")) { statement =>
+    using(sqlite.prepareStatement("SELECT * FROM relayed WHERE timestamp >= ? AND timestamp < ?")) { statement =>
       statement.setLong(1, from)
       statement.setLong(2, to)
       val rs = statement.executeQuery()
-      var q: Queue[PaymentRelayed] = Queue()
+      var relayedByHash = Map.empty[ByteVector32, Seq[RelayedPart]]
       while (rs.next()) {
-        q = q :+ ChannelPaymentRelayed(
-          amountIn = MilliSatoshi(rs.getLong("amount_in_msat")),
-          amountOut = MilliSatoshi(rs.getLong("amount_out_msat")),
-          paymentHash = rs.getByteVector32("payment_hash"),
-          fromChannelId = rs.getByteVector32("from_channel_id"),
-          toChannelId = rs.getByteVector32("to_channel_id"),
-          timestamp = rs.getLong("timestamp"))
+        val paymentHash = rs.getByteVector32("payment_hash")
+        val part = RelayedPart(
+          rs.getByteVector32("channel_id"),
+          MilliSatoshi(rs.getLong("amount_msat")),
+          rs.getString("direction"),
+          rs.getString("relay_type"),
+          rs.getLong("timestamp"))
+        relayedByHash = relayedByHash + (paymentHash -> (relayedByHash.getOrElse(paymentHash, Nil) :+ part))
```

Comment:
There is also this pattern:

```scala
var relayedByHash = Map.empty[ByteVector32, Seq[RelayedPart]].withDefaultValue(Nil)
...
relayedByHash = relayedByHash + (paymentHash -> (relayedByHash(paymentHash) :+ part))
```

But I'm not sure I like it better

Reply:
I don't have a strong opinion, whatever you feel is better!

Reply:
Let's leave it like that